drivers/pwm/pwm_rcar.c

```diff
@@ -0,0 +1,279 @@
+}
+
+static void pwm_rcar_write_bit(const struct pwm_rcar_cfg *config,
+			       uint32_t offs, uint32_t bits, bool value)
+{
+	uint32_t reg_val = pwm_rcar_read(config, offs);
+
+	if (value) {
+		reg_val |= bits;
+	} else {
+		reg_val &= ~(bits);
+	}
+
+	pwm_rcar_write(config, offs, reg_val);
+}
+
+static int pwm_rcar_update_clk(const struct pwm_rcar_cfg *config, uint32_t pwm,
+			       uint32_t *period_cycles, uint32_t *pulse_cycles)
+{
+	uint32_t reg_val, power, diviser;
+
+	power = pwm_rcar_read(config, RCAR_PWM_CR) & RCAR_PWM_DIVISER_MASK;
+	power = power >> RCAR_PWM_DIVISER_SHIFT;
+	diviser = 1 << power;
+
+	LOG_DBG("Found old diviser : 2^%d=%d", power, diviser);
+
+	/* Looking for the best possible clock diviser */
+	if (*period_cycles > RCAR_PWM_MAX_CYCLE) {
+		/* Reducing clock speed */
+		while (*period_cycles > RCAR_PWM_MAX_CYCLE) {
+			diviser *= 2;
+			*period_cycles /= 2;
+			*pulse_cycles /= 2;
+			power++;
+			if (power > RCAR_PWM_MAX_DIV) {
+				return -ENOTSUP;
+			}
+		}
+	} else {
+		/* Increasing clock speed */
+		while (*period_cycles < (RCAR_PWM_MAX_CYCLE / 2)) {
+			if (power == 0) {
+				return -ENOTSUP;
+			}
+			diviser /= 2;
+			*period_cycles *= 2;
+			*pulse_cycles *= 2;
+			power--;
+		}
+	}
+	LOG_DBG("Found new diviser : 2^%d=%d\n", power, diviser);
+
+	/* Set new clock Diviser */
+	reg_val = pwm_rcar_read(config, RCAR_PWM_CR);
+	reg_val &= ~RCAR_PWM_DIVISER_MASK;
+	reg_val |= (power << RCAR_PWM_DIVISER_SHIFT);
+	pwm_rcar_write(config, RCAR_PWM_CR, reg_val);
+
+	return 0;
+}
+
+static int pwm_rcar_pin_set(const struct device *dev, uint32_t pwm,
+			    uint32_t period_cycles, uint32_t pulse_cycles,
+			    pwm_flags_t flags)
+{
+	const struct pwm_rcar_cfg *config = dev->config;
+	uint32_t reg_val;
+	int ret = 0;
+
+	/* Assume only one channel per pwm instance */
+	if (pwm != 0) {
+		return -ENOTSUP;
+	}
+
+	if (flags != PWM_POLARITY_NORMAL) {
+		return -ENOTSUP;
+	}
+
+	/* Prohibiten values */
+	if (period_cycles == 0U || pulse_cycles == 0U ||
+	    pulse_cycles > period_cycles) {
+		return -EINVAL;
+	}
+
+	LOG_DBG("base_reg=0x%x, pulse_cycles=%d, period_cycles=%d,"
+		" duty_cycle=%d", config->reg_addr, pulse_cycles,
+		period_cycles, (pulse_cycles * 100U / period_cycles));
+
+	/* Disable PWM */
+	pwm_rcar_write_bit(config, RCAR_PWM_CR, RCAR_PWM_CR_EN, false);
+
+	/* Set continuous mode */
+	pwm_rcar_write_bit(config, RCAR_PWM_CR, RCAR_PWM_CR_SS, false);
+
+	/* Enable SYNC mode */
+	pwm_rcar_write_bit(config, RCAR_PWM_CR, RCAR_PWM_CR_SYNC, true);
+
+	/*
+	 * Set clock counter according to the requested period_cycles
+	 * if period_cycles is less than half of the counter, then the
+	 * clock diviser could be updated as the diviser is a modulo 2.
+	 */
+	if (period_cycles > RCAR_PWM_MAX_CYCLE ||
+	    period_cycles < (RCAR_PWM_MAX_CYCLE / 2)) {
+		LOG_DBG("Adapting frequency diviser...");
+		ret = pwm_rcar_update_clk(config, pwm, &period_cycles, &pulse_cycles);
+		if (ret != 0) {
+			return ret;
+		}
+	}
```

Comment:
Spaces and tabulation are mixed here, please unify this in the driver file.

Reply:
Theses spaces are made by uncrustify to align the "if" arguments.